pkg/collect/certificates.go

```diff
@@ -0,0 +1,230 @@
+package collect
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"encoding/pem"
+	"time"
+
+	troubleshootv1beta2 "github.com/replicatedhq/troubleshoot/pkg/apis/troubleshoot/v1beta2"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+type CollectCertificates struct {
+	Collector    *troubleshootv1beta2.Certificates
+	BundlePath   string
+	Namespace    string
+	ClientConfig *rest.Config
+	Client       kubernetes.Interface
+	Context      context.Context
+	RBACErrors
+}
+
+// Certificate collection struct
+type CertCollection struct {
+	Source           *CertificateSource  `json:"source"`
+	Errors           []string            `json:"errors"`
+	CertificateChain []ParsedCertificate `json:"certificateChain"`
+}
+
+// Certificate source
+type CertificateSource struct {
+	SecretName    string `json:"secret,omitempty"`
+	ConfigMapName string `json:"configMap,omitempty"`
+	Namespace     string `json:"namespace,omitempty"`
+}
+
+// Certificate Struct
+type ParsedCertificate struct {
+	CertName                string    `json:"certificate"`
+	Subject                 string    `json:"subject"`
+	SubjectAlternativeNames []string  `json:"subjectAlternativeNames"`
+	Issuer                  string    `json:"issuer"`
+	NotAfter                time.Time `json:"notAfter"`
+	NotBefore               time.Time `json:"notBefore"`
+	IsValid                 bool      `json:"isValid"`
+	IsCA                    bool      `json:"isCA"`
+}
+
+func (c *CollectCertificates) Title() string {
+	return getCollectorName(c)
+}
+
+func (c *CollectCertificates) IsExcluded() (bool, error) {
+	return isExcluded(c.Collector.Exclude)
+}
+
+func (c *CollectCertificates) Collect(progressChan chan<- interface{}) (CollectorResult, error) {
+
+	output := NewResult()
+	results := []CertCollection{}
+
+	// collect certificates from secrets
+	for secretName, namespaces := range c.Collector.Secrets {
+		for _, namespace := range namespaces {
+
+			secretCollections := secretCertCollector(secretName, namespace, c.Client)
+			results = append(results, secretCollections...)
+		}
+	}
+
+	// collect certificates from configMaps
+	for configMapName, namespaces := range c.Collector.ConfigMaps {
+		for _, namespace := range namespaces {
+			configMapCollections := configMapCertCollector(configMapName, namespace, c.Client)
+			results = append(results, configMapCollections...)
+		}
+	}
+
+	certsJson, _ := json.MarshalIndent(results, "", "\t")
+
+	filePath := "certificates/certificates.json"
+
+	output.SaveResult(c.BundlePath, filePath, bytes.NewBuffer(certsJson))
```

Comment:
Could you please also add error handling here

```go
err := output.SaveResult(c.BundlePath, filePath, bytes.NewBuffer(certsJson))
if err != nil {
	return nil, err
}
```

Reply:
completed